File: 0509.斐波那契数/0509-斐波那契数.py
# from functools import wraps
class Solution:
    def fib(self, N: int) -> int:
        def simple_cache(fn):
            d = {}
            # @wraps
            def calc(n):
                if n in d:
                    return d[n]
                d[n] = fn(n)
                return d[n]
            return calc
        
        @simple_cache
        def fib(n):
            if n < 2:
                return n
            return fib(n-1) + fib(n-2)
        return fib(N)
        

File: 0509.斐波那契数/0509-斐波那契数.rs
impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n < 2 {
            return n;
        }
        let mut n = n;
        let mut a = 0;
        let mut b = 1;
        while n > 1 {
            let tmp = b;
            b = a +b;
            a = tmp;
            n -= 1;
        }
        return b;
        /*
        match n {
            0 => 0,
            1 => 1,
            _ => Self::fib(n-1) + Self::fib(n-2),
        }
        */
    }
}